src/mds/CInode.cc

```diff
@@ -4383,3 +4390,126 @@ int64_t CInode::get_backtrace_pool() const
     return inode.layout.pool_id;
   }
 }
+
+class C_CInode_ExportPin : public MDSInternalContext {
+public:
+  explicit C_CInode_ExportPin(CInode *in) : MDSInternalContext(in->mdcache->mds), in(in) {
+    in->get(MDSCacheObject::PIN_PTRWAITER);
+  }
+  ~C_CInode_ExportPin() {
+    in->put(MDSCacheObject::PIN_PTRWAITER);
+  }
+
+  void finish(int r) override {
+    in->maybe_export_pin();
+  }
+private:
+  CInode *in;
+};
+
+void CInode::maybe_export_pin()
+{
+  if (g_conf->mds_bal_export_pin && is_dir()) {
+    dout(20) << "maybe_export_pin " << *this << dendl;
+    mds_rank_t pin = get_projected_inode()->export_pin;
+    if (pin == MDS_RANK_NONE) {
+      if (!mdcache->export_pin_queue.count(this)) {
+        /* Try to find farthest full auth parent fragment which is pinned
+         * elsewhere.  There cannot be a break in the authority chain of
+         * directories, otherwise this inode itself will not be exported.
+         */
+        CInode *auth_last = this; /* N.B. we may not be full auth for any fragments of this inode, but adding it to the queue is harmless. */
+        bool auth_barrier = false;
+        for (CDir *cd = get_projected_parent_dir(); cd && !cd->inode->is_base() && !cd->inode->is_system(); cd = cd->inode->get_projected_parent_dir()) {
+          if (cd->is_full_dir_auth() && !auth_barrier) {
+            auth_last = cd->inode;
+          } else {
+            auth_barrier = true;
+          }
+          pin = cd->inode->get_projected_inode()->export_pin;
+          if (pin != MDS_RANK_NONE) {
+            if (pin != mdcache->mds->get_nodeid()) {
+              dout(20) << "adding ancestor to export_pin_queue " << *auth_last << dendl;
+              mdcache->export_pin_queue.insert(auth_last);
+            }
+            break;
+          }
+        }
+      }
+    } else {
+      if (pin == mdcache->mds->get_nodeid()) {
+        for (auto it = dirfrags.begin(); it != dirfrags.end(); it++) {
+          CDir *cd = it->second;
+          if (cd->state_test(CDir::STATE_AUXSUBTREE)) continue;
+          CDir *subtree = mdcache->get_subtree_root(cd);
+          if (!subtree) continue;
+          if (subtree->is_ambiguous_auth()) {
+            subtree->add_waiter(MDSCacheObject::WAIT_SINGLEAUTH, new C_CInode_ExportPin(this));
+            dout(15) << "aux subtree pin of " << *cd << " delayed for single auth on subtree " << *subtree << dendl;
+          } else if (subtree->is_auth()) {
+            assert(cd->is_auth());
+            if (subtree->is_frozen() || subtree->is_freezing()) {
+              subtree->add_waiter(MDSCacheObject::WAIT_UNFREEZE, new C_CInode_ExportPin(this));
+              dout(15) << "aux subtree pin of " << *cd << " delayed for unfreeze on subtree " << *subtree << dendl;
+            } else {
+              cd->state_set(CDir::STATE_AUXSUBTREE);
```

Comment:
If we can remove above code, we can clean this function further. move most of the code into MDBalancer::handle_export_pins(). this function just add inode to mdcache->export_pin_queue